crates/sui-core/src/authority_server.rs

```diff
@@ -540,6 +542,198 @@ impl ValidatorService {
             fastpath_input_objects: vec![], // unused field
         }))
     }
+
+    // Handle_soft_bundle_certificates submits a batch of certificates.
+    // If an error occurs for any of the certificates, the error is returned and the remaining certificates are not processed.
+    // Effects are always returned.
+    async fn handle_soft_bundle_certificates(
+        self,
+        certificates: Vec<CertifiedTransaction>,
+    ) -> Result<HandleSoftBundleCertificatesResponseV2, tonic::Status> {
+        let Self {
+            state,
+            consensus_adapter,
+            metrics,
+            traffic_controller: _,
+        } = self;
+
+        let epoch_store = state.load_epoch_store_one_call_per_task();
+        // Validate if cert can be executed
+        // Fullnode does not serve handle_certificate call.
+        fp_ensure!(
+            !state.is_fullnode(&epoch_store),
+            SuiError::FullNodeCantHandleCertificate.into()
+        );
+        let protocol_config = epoch_store.protocol_config();
+        fp_ensure!(
+            protocol_config.enable_soft_bundle(),
+            SuiError::FailedToSubmitToConsensus("SoftBundle: Feature not enabled".into()).into()
+        );
+        let max_num_certs = protocol_config.soft_bundle_max_num_certificates() as usize;
+        let max_serialized_size = protocol_config.soft_bundle_max_serialized_size_bytes() as usize;
+        fp_ensure!(
+            certificates.len() > 0,
+            SuiError::FailedToSubmitToConsensus("SoftBundle: Bundle size must not be zero".into())
+                .into()
+        );
+        fp_ensure!(
+            certificates.len() <= max_num_certs,
+            SuiError::FailedToSubmitToConsensus(
+                "SoftBundle: Bundle contains too many certificates".into()
```

Comment:
For the two preceding errors, can you use SuiError::UserInputError and add a new UserInputError variant for InvalidBundleSize or something like that?